src/org.hdfgroup.object/hdf/object/Attribute.java

```diff
@@ -14,447 +14,38 @@
-     * applies to both HDF4 and HDF5.
-     * <p>
-     * The following example creates a string attribute with the name "CLASS" and
-     * value "IMAGE".
-     *
-     * <pre>
-     * long[] attrDims = { 1 };
-     * String attrName = &quot;CLASS&quot;;
-     * String[] classValue = { &quot;IMAGE&quot; };
-     * Datatype attrType = null;
-     * try {
-     *     attrType = new H5Datatype(Datatype.CLASS_STRING, classValue[0].length() + 1, Datatype.NATIVE, Datatype.NATIVE);
-     * }
-     * catch (Exception ex) {}
-     * Attribute attr = new Attribute(attrName, attrType, attrDims, classValue);
-     * </pre>
-     *
-     * @param parentObj
-     *            the HObject to which this Attribute is attached.
-     * @param attrName
-     *            the name of the attribute.
-     * @param attrType
-     *            the datatype of the attribute.
-     * @param attrDims
-     *            the dimension sizes of the attribute, null for scalar attribute
-     * @param attrValue
-     *            the value of the attribute, null if no value
-     *
-     * @see hdf.object.Datatype
-     */
-    @SuppressWarnings({ "rawtypes", "unchecked", "deprecation" })
-    public Attribute(HObject parentObj, String attrName, Datatype attrType, long[] attrDims, Object attrValue) {
-        super((parentObj == null) ? null : parentObj.getFileFormat(), attrName,
-                (parentObj == null) ? null : parentObj.getFullName(), null);
-
-        log.trace("Attribute: start {}", parentObj);
-        this.parentObject = parentObj;
-
-        datatype = attrType;
-        if (attrDims == null) {
-            rank = 1;
-            dims = new long[] { 1 };
-            isScalar = true;
-        }
-        else {
-            dims = attrDims;
-            rank = dims.length;
-            isScalar = false;
-        }
-
-        data = attrValue;
-        properties = new HashMap();
-
-        unsignedConverted = false;
-
-        selectedDims = new long[rank];
-        startDims = new long[rank];
-        selectedStride = new long[rank];
-
-        log.trace("attrName={}, attrType={}, attrValue={}, rank={}, isUnsigned={}, isScalar={}",
-                attrName, getDatatype().getDescription(), data, rank, getDatatype().isUnsigned(), isScalar);
-
-        resetSelection();
-    }
-
-    /*
-     * (non-Javadoc)
-     *
-     * @see hdf.object.HObject#open()
-     */
-    @Override
-    public long open() {
-        if (parentObject == null) {
-            log.debug("open(): attribute's parent object is null");
-            return -1;
-        }
-
-        return -1;
-    }
-
-    /*
-     * (non-Javadoc)
-     *
-     * @see hdf.object.HObject#close(int)
-     */
-    @Override
-    public void close(long aid) {
-    }
-
-    @Override
-    public void init() {
-        if (inited) {
-            resetSelection();
-            log.trace("init(): Attribute already inited");
-            return;
-        }
-    }
+public interface Attribute {
```

Comment:
I honestly don't think that going to an interface for Attributes is the right move here. By going to an interface, you've now changed all the essential methods (getDatatype, getName(), getData()) that were inherited from the DataFormat and/or CompoundDataFormat interfaces into ones that are specialized for Attributes. This will make it much more complicated to write code that works across object types within classes like TableViews, for example. Scalar attributes have been functioning exactly like datasets for a while now since we rewrote the two classes to share the common DataFormat interface. So, theoretically it should be the same for compound attributes and datasets; the CompoundDataFormat interface may just need a little reworking.